File: Stripe/STPSetupIntentConfirmParams.m
//
//  STPSetupIntentConfirmParams.m
//  Stripe
//
//  Created by Yuki Tokuhiro on 6/27/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

#import "STPSetupIntentConfirmParams.h"
#import "STPSetupIntentConfirmParams+Utilities.h"

#import "STPMandateCustomerAcceptanceParams.h"
#import "STPMandateOnlineParams+Private.h"
#import "STPMandateDataParams.h"
#import "STPPaymentMethodParams.h"

@interface STPSetupIntentConfirmParams ()

@end

@implementation STPSetupIntentConfirmParams

@synthesize additionalAPIParameters = _additionalAPIParameters;

- (instancetype)init {
    // Not a valid clientSecret, but at least it'll be non-null
    return [self initWithClientSecret:@""];
}

- (instancetype)initWithClientSecret:(NSString *)clientSecret {
    self = [super init];
    if (self) {
        _clientSecret = [clientSecret copy];
        _additionalAPIParameters = @{};
    }
    return self;
}

- (NSString *)description {
    NSArray *props = @[
                       // Object
                       [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],
                       
                       // SetupIntentParams details (alphabetical)
                       [NSString stringWithFormat:@"clientSecret = %@", (self.clientSecret.length > 0) ? @"<redacted>" : @""],
                       [NSString stringWithFormat:@"returnURL = %@", self.returnURL],
                       [NSString stringWithFormat:@"paymentMethodId = %@", self.paymentMethodID],
                       [NSString stringWithFormat:@"paymentMethodParams = %@", self.paymentMethodParams],
                       [NSString stringWithFormat:@"useStripeSDK = %@", self.useStripeSDK],

                       // Mandate
                       [NSString stringWithFormat:@"mandateData = %@", self.mandateData],
                       [NSString stringWithFormat:@"mandate = %@", self.mandate],

                       // Additional params set by app
                       [NSString stringWithFormat:@"additionalAPIParameters = %@", self.additionalAPIParameters],
                       ];
    
    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

- (STPMandateDataParams *)mandateData {
    BOOL paymentMethodRequiresMandate = self.paymentMethodParams.type == STPPaymentMethodTypeSEPADebit || self.paymentMethodParams.type == STPPaymentMethodTypeBacsDebit || self.paymentMethodParams.type == STPPaymentMethodTypeAUBECSDebit;
    
    if (_mandateData != nil) {
        return _mandateData;
    } else if (self.mandate == nil && paymentMethodRequiresMandate) {
        // Create default infer from client mandate_data
        STPMandateDataParams *mandateData = [[STPMandateDataParams alloc] init];
        STPMandateCustomerAcceptanceParams *customerAcceptance = [[STPMandateCustomerAcceptanceParams alloc] init];
        STPMandateOnlineParams *onlineParams = [[STPMandateOnlineParams alloc] init];
        onlineParams.inferFromClient = @YES;
        customerAcceptance.type = STPMandateCustomerAcceptanceTypeOnline;
        customerAcceptance.onlineParams = onlineParams;
        mandateData.customerAcceptance = customerAcceptance;

        return mandateData;
    } else {
        return nil;
    }
}

#pragma mark - NSCopying

- (instancetype)copyWithZone:(NSZone *)zone {
    __typeof(self) copy = [[[self class] allocWithZone:zone] init];

    copy.clientSecret = self.clientSecret;
    copy.paymentMethodParams = self.paymentMethodParams;
    copy.paymentMethodID = self.paymentMethodID;
    copy.returnURL = self.returnURL;
    copy.useStripeSDK = self.useStripeSDK;
    copy.mandateData = self.mandateData;
    copy.mandate = self.mandate;
    copy.additionalAPIParameters = self.additionalAPIParameters;

    return copy;
}

#pragma mark - STPFormEncodable

+ (nullable NSString *)rootObjectName {
    return nil;
}

+ (nonnull NSDictionary *)propertyNamesToFormFieldNamesMapping {
    return @{
             NSStringFromSelector(@selector(clientSecret)): @"client_secret",
             NSStringFromSelector(@selector(paymentMethodParams)): @"payment_method_data",
             NSStringFromSelector(@selector(paymentMethodID)): @"payment_method",
             NSStringFromSelector(@selector(returnURL)): @"return_url",
             NSStringFromSelector(@selector(useStripeSDK)): @"use_stripe_sdk",
             NSStringFromSelector(@selector(mandateData)): @"mandate_data",
             NSStringFromSelector(@selector(mandateData)) : @"mandate_data",
             NSStringFromSelector(@selector(mandate)) : @"mandate",
             };
}

#pragma mark - Utilities

+ (BOOL)isClientSecretValid:(NSString *)clientSecret {
    static dispatch_once_t onceToken;
    static NSRegularExpression *regex = nil;
    dispatch_once(&onceToken, ^{
        regex = [[NSRegularExpression alloc] initWithPattern:@"^seti_[^_]+_secret_[^_]+$"
                                                     options:0
                                                       error:NULL];
    });

    return ([regex numberOfMatchesInString:clientSecret
                                   options:NSMatchingAnchored
                                     range:NSMakeRange(0, clientSecret.length)]) == 1;
}

@end

File: Stripe/STPSetupIntentConfirmParams.swift
//
//  STPSetupIntentConfirmParams.swift
//  Stripe
//
//  Created by Yuki Tokuhiro on 6/27/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

import Foundation

/// An object representing parameters to confirm a SetupIntent object.
/// For example, you would confirm a SetupIntent when a customer hits the “Save” button on a payment method management view in your app.
/// If the selected payment method does not require any additional steps from the customer, the SetupIntent's status will transition to `STPSetupIntentStatusSucceeded`.  Otherwise, it will transition to `STPSetupIntentStatusRequiresAction`, and suggest additional actions via `nextAction`.
/// Instead of passing this to `STPAPIClient.confirmSetupIntent(...)` directly, we recommend using `STPPaymentHandler` to handle any additional steps for you.
/// - seealso: https://stripe.com/docs/api/setup_intents/confirm
public class STPSetupIntentConfirmParams: NSObject, NSCopying, STPFormEncodable {
  @objc public var additionalAPIParameters: [AnyHashable: Any] = [:]

  /// Initialize this `STPSetupIntentConfirmParams` with a `clientSecret`.
  /// - Parameter clientSecret: the client secret for this SetupIntent
  @objc
  public init(clientSecret: String) {
    self.clientSecret = clientSecret
    super.init()
    additionalAPIParameters = [:]
  }

  /// The client secret of the SetupIntent. Required.
  @objc public var clientSecret: String
  /// Provide a supported `STPPaymentMethodParams` object, and Stripe will create a
  /// PaymentMethod during PaymentIntent confirmation.
  /// @note alternative to `paymentMethodId`
  @objc public var paymentMethodParams: STPPaymentMethodParams?
  /// Provide an already created PaymentMethod's id, and it will be used to confirm the SetupIntent.
  /// @note alternative to `paymentMethodParams`
  @objc public var paymentMethodID: String?
  /// The URL to redirect your customer back to after they authenticate or cancel
  /// their payment on the payment method’s app or site.
  /// This should probably be a URL that opens your iOS app.
  @objc public var returnURL: String?
  /// A boolean number to indicate whether you intend to use the Stripe SDK's functionality to handle any SetupIntent next actions.
  /// If set to false, STPSetupIntent.nextAction will only ever contain a redirect url that can be opened in a webview or mobile browser.
  /// When set to true, the nextAction may contain information that the Stripe SDK can use to perform native authentication within your
  /// app.
  @objc public var useStripeSDK: NSNumber?
  /// Details about the Mandate to create.
  /// @note If this value is null and the `(self.paymentMethod.type == STPPaymentMethodTypeSEPADebit | | self.paymentMethodParams.type == STPPaymentMethodTypeAUBECSDebit || self.paymentMethodParams.type == STPPaymentMethodTypeBacsDebit) && self.mandate == nil`, the SDK will set this to an internal value indicating that the mandate data should be inferred from the current context.
  @objc public var mandateData: STPMandateDataParams? {
    set(newMandateData) {
      _mandateData = newMandateData
    }
    get {
      if let _mandateData = _mandateData {
        return _mandateData
      }

      if let paymentMethodParams = self.paymentMethodParams,
        paymentMethodParams.type == .SEPADebit || paymentMethodParams.type == .bacsDebit
          || paymentMethodParams.type == .AUBECSDebit
      {

        // Create default infer from client mandate_data
        let onlineParams = STPMandateOnlineParams(ipAddress: "", userAgent: "")
        onlineParams.inferFromClient = NSNumber(value: true)
        let customerAcceptance = STPMandateCustomerAcceptanceParams()
        customerAcceptance.type = .online
        customerAcceptance.onlineParams = onlineParams
        let mandateData = STPMandateDataParams(customerAcceptance: customerAcceptance)
        return mandateData
      } else {
        return nil
      }
    }
  }
  private var _mandateData: STPMandateDataParams?

  override convenience init() {
    // Not a valid clientSecret, but at least it'll be non-null
    self.init(clientSecret: "")
  }

  /// :nodoc:
  @objc public override var description: String {
    let props = [
      // Object
      String(format: "%@: %p", NSStringFromClass(STPSetupIntentConfirmParams.self), self),
      // SetupIntentParams details (alphabetical)
      "clientSecret = \(((clientSecret.count) > 0) ? "<redacted>" : "")",
      "returnURL = \(returnURL ?? "")",
      "paymentMethodId = \(paymentMethodID ?? "")",
      "paymentMethodParams = \(String(describing: paymentMethodParams))",
      "useStripeSDK = \(useStripeSDK ?? 0)",
      // Mandate
      "mandateData = \(String(describing: mandateData))",
      // Additional params set by app
      "additionalAPIParameters = \(additionalAPIParameters )",
    ]

    return "<\(props.joined(separator: "; "))>"
  }

  // MARK: - NSCopying
  /// :nodoc:
  @objc
  public func copy(with zone: NSZone? = nil) -> Any {
    let copy = STPSetupIntentConfirmParams()

    copy.clientSecret = clientSecret
    copy.paymentMethodParams = paymentMethodParams
    copy.paymentMethodID = paymentMethodID
    copy.returnURL = returnURL
    copy.useStripeSDK = useStripeSDK
    copy.mandateData = mandateData
    copy.additionalAPIParameters = additionalAPIParameters

    return copy
  }

  // MARK: - STPFormEncodable
  public class func rootObjectName() -> String? {
    return nil
  }

  public class func propertyNamesToFormFieldNamesMapping() -> [String: String] {
    return [
      NSStringFromSelector(#selector(getter:clientSecret)): "client_secret",
      NSStringFromSelector(#selector(getter:paymentMethodParams)): "payment_method_data",
      NSStringFromSelector(#selector(getter:paymentMethodID)): "payment_method",
      NSStringFromSelector(#selector(getter:returnURL)): "return_url",
      NSStringFromSelector(#selector(getter:useStripeSDK)): "use_stripe_sdk",
      NSStringFromSelector(#selector(getter:mandateData)): "mandate_data",
    ]
  }

  // MARK: - Utilities
  static private let regex = try! NSRegularExpression(
    pattern: "^seti_[^_]+_secret_[^_]+$", options: [])
  class func isClientSecretValid(_ clientSecret: String) -> Bool {
    return
      (regex.numberOfMatches(
        in: clientSecret,
        options: .anchored,
        range: NSRange(location: 0, length: clientSecret.count))) == 1
  }
}
